nano/node/messages.cpp

```diff
@@ -1581,4 +1604,151 @@ std::size_t nano::node_id_handshake::size (nano::message_header const & header_a
 		result += sizeof (nano::account) + sizeof (nano::signature);
 	}
 	return result;
+}
+
+/*
+ * asc_pull_req
+ */
+
+nano::asc_pull_req::asc_pull_req (const nano::network_constants & constants) :
+	message (constants, nano::message_type::asc_pull_req)
+{
+}
+
+nano::asc_pull_req::asc_pull_req (bool & error, nano::stream & stream, const nano::message_header & header) :
+	message (header)
+{
+	error = deserialize (stream);
+}
+
+void nano::asc_pull_req::visit (nano::message_visitor & visitor) const
+{
+	visitor.asc_pull_req (*this);
+}
+
+void nano::asc_pull_req::serialize (nano::stream & stream) const
+{
+	header.serialize (stream);
+	nano::write (stream, type);
+	nano::write (stream, id);
+	nano::write (stream, start);
+}
+
+bool nano::asc_pull_req::deserialize (nano::stream & stream)
+{
+	debug_assert (header.type == nano::message_type::asc_pull_req);
+	bool error = false;
+	try
+	{
+		nano::read (stream, type);
+		nano::read (stream, id);
+		nano::read (stream, start);
+	}
+	catch (std::runtime_error const &)
+	{
+		error = true;
+	}
+	return error;
+}
+
+/*
+ * asc_pull_ack
+ */
+
+nano::asc_pull_ack::asc_pull_ack (const nano::network_constants & constants) :
+	message (constants, nano::message_type::asc_pull_ack)
+{
+	update_header ();
+}
+
+nano::asc_pull_ack::asc_pull_ack (bool & error, nano::stream & stream, const nano::message_header & header) :
+	message (header)
+{
+	error = deserialize (stream);
+	if (!error)
+	{
+		update_header ();
```

Comment:
Why is this updating the header? The header should either be right or wrong. It should not need updating/adjusting.

Reply:
Good catch, there is no reason to do that in there. Personally I would update the header during serialization, but that is the const method and I didn't want to suddenly refactor all message classes